pkg/querier/queryrange/querysharding.go

```diff
@@ -0,0 +1,311 @@
+package queryrange
+
+import (
+	"context"
+	fmt "fmt"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/querier/astmapper"
+	"github.com/cortexproject/cortex/pkg/querier/lazyquery"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/promql"
+)
+
+var (
+	nanosecondsInMillisecond = int64(time.Millisecond / time.Nanosecond)
+
+	errInvalidShardingRange = errors.New("Query does not fit in a single sharding configuration")
+)
+
+// ShardingConfigs is a slice of chunk shard configs
+type ShardingConfigs []chunk.PeriodConfig
+
+// ValidRange extracts a non-overlapping sharding configuration from a list of configs and a time range.
+func (confs ShardingConfigs) ValidRange(start, end int64) (chunk.PeriodConfig, error) {
+	for i, conf := range confs {
+		if start < int64(conf.From.Time) {
+			// the query starts before this config's range
+			return chunk.PeriodConfig{}, errInvalidShardingRange
+		} else if i == len(confs)-1 {
+			// the last configuration has no upper bound
+			return conf, nil
+		} else if end < int64(confs[i+1].From.Time) {
+			// The request is entirely scoped into this shard config
+			return conf, nil
+		} else {
+			continue
+		}
+	}
+
+	return chunk.PeriodConfig{}, errInvalidShardingRange
+}
+
+func (confs ShardingConfigs) hasShards() bool {
+	for _, conf := range confs {
+		if conf.RowShards > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func mapQuery(mapper astmapper.ASTMapper, query string) (promql.Node, error) {
+	expr, err := promql.ParseExpr(query)
+	if err != nil {
+		return nil, err
+	}
+	return mapper.Map(expr)
+}
+
+// NewQueryShardMiddleware creates a middleware which downstreams queries after AST mapping and query encoding.
+func NewQueryShardMiddleware(
+	logger log.Logger,
+	engine *promql.Engine,
+	confs ShardingConfigs,
+	codec Codec,
+	minShardingLookback time.Duration,
+) Middleware {
+	passthrough := MiddlewareFunc(func(next Handler) Handler {
+		return next
+	})
+
+	noshards := !confs.hasShards()
+
+	if noshards {
+		level.Warn(logger).Log(
+			"middleware", "QueryShard",
+			"msg", "no configuration with shard found",
+			"confs", fmt.Sprintf("%+v", confs),
+		)
+		return passthrough
+	}
+
+	getConf := func(r Request) (chunk.PeriodConfig, error) {
```

Comment:
I think this could be a function of the alias `ShardingConfigs` created line 24. This should simplify the code by using the alias as dependency instead of a unnamed function that you had to defined 3 time.

Reply:
great idea!